### PR TITLE
Allow searching for phone numbers with parenthesis, spaces and dashes in them

### PR DESCRIPTION
**Pull request by narefyev91**
### Details
The main issue here that search mechanism is not understand formatted phones. We need to catch when user typing/inserting phone number and prepare it in correct format to make searching mechanism working fine

We will use already existing regExp - which working fine and covers all our needs from `CONST` folder.
Also to make it working in any other places we need to apply changes for `getOptions` function instead of fixing this issue around different places.
`getOptions` function has more that 41 variable with name `searchValue` inside itself - and to make less changes in the whole function i suggest to rename prop from `searchValue` to `searchInputValue` and inside `getOptions` define a variable `searchValue` and assign formatted value if needed. Also to not break existing functionality (like this one
```
 const login = (Str.isValidPhone(searchValue) && !searchValue.includes('+'))
        ? `+${countryCodeByIP}${searchValue}`
        : searchValue;
```
 - we will remove any characters which are **not numbers but will allow to have +**

### Fixed Issues

$ https://github.com/Expensify/App/issues/15491
$ https://github.com/Expensify/App/issues/15491#issuecomment-1448125134

### Tests
1. Open search
2. Type phone/paste phone number in this formats (xxx) xxx-xxxx / (xxx)xxx-xxxx / +x(xxx)xxx-xxxx / +x (xxx) xxx-xxxx
3. Verify that search return correct result

- [ ] Verify that no errors appear in the JS console

### Offline tests
This feature doesn't change or is impacted by offline mode.

### QA Steps
Same as above.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/28352309/222187526-b4956f70-7b6b-4c83-86a0-162d350ece72.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/28352309/222187680-8bf4c04b-dbe3-4a77-a6b1-60db13b93a30.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/28352309/222187610-e4c4f51a-b01b-4f0a-823d-8c62559f4cc7.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/28352309/222187716-5d49c592-7970-4338-90e7-927ef34c6d53.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/28352309/222187749-a63824d5-57b7-4a17-b0e6-0c7994b9eb54.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/28352309/222187790-afdf0fd9-b74e-4354-b602-333001274b88.mov

</details>

### CODE REVIEW

**Comment by puneetlath** [comment]
I think it's close but not quite there. 

It accepts `(972)658-5619`

image

But not `+1(972)658-5619`

image

And not `+1 (972) 658-5619`

image

It should accept all of those.

**Comment by narefyev91** [comment]
@puneetlath done
Simulator Screen Shot - iPhone 14 Pro - 2023-03-01 at 18 18 51
Simulator Screen Shot - iPhone 14 Pro - 2023-03-01 at 18 18 38
Simulator Screen Shot - iPhone 14 Pro - 2023-03-01 at 18 18 17
Simulator Screen Shot - iPhone 14 Pro - 2023-03-01 at 18 18 01

File: src/CONST.js [hunk 1]
```diff
@@ -786,6 +786,7 @@ const CONST = {
         DIGITS_AND_PLUS: /^\+?[0-9]*$/,
         PHONE_E164_PLUS: /^\+?[1-9]\d{1,14}$/,
         PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]{0,12}$/,
+        PHONE_WITH_ANY_CHARS: /^\s*(?:\+?(\d{1,3}))?[-. (]*(\d{3})[-. )]*(\d{3})[-. ]*(\d{4})(?: *x(\d+))?\s*$/,
```
Comment (puneetlath) [inline src/CONST.js:789]: Could you describe what this regex is doing and how it is different than `PHONE_WITH_SPECIAL_CHARS`? I just want to make sure I understand it.

**Comment by puneetlath** [comment]
@sobitneupane can you please give this a review and run it through the reviewer checklist? Thanks!

File: src/CONST.js (hunk 1 above)
Comment (narefyev91) [reply src/CONST.js:789]: @puneetlath the main difference coming to this line `(?:\+?(\d{1,3}))` we check starting element may have + or number before (). The `PHONE_WITH_SPECIAL_CHARS` does not support that - and this line +1(800)/ 1(800) will not be catch as phone number

**Comment by sobitneupane** [comment]
@puneetlath What are the formats to be accepted? I have list few based on above discussions:

- (972)658-5619
- +1(972)658-5619
- +1 (972) 658-5619

Other countries might have different formats. Should we allow special characters `(` , `)` `-` and space ` ` in all positons?

**Comment by puneetlath** [comment]
Good question. I think definitely those three that you listed. Are there known phone number formats that this doesn't support?

**Comment by sobitneupane** [comment]
@puneetlath

I have done some internet surfing. I am not sure how much it makes sense in real world.

Source: https://blog.insycle.com/phone-number-formatting-crm
Screenshot 2023-03-02 at 22 34 50

001-212-456-7890
191-212-456-7890

Source: https://en.wikipedia.org/wiki/National_conventions_for_writing_telephone_numbers

France: +49 AAAA BBBBBB
NewZeland: +64-MM-BBB-BBBB
India +91-AAAAA BBBBB

These are few. Each country has different format.

**Comment by narefyev91** [comment]
@sobitneupane @puneetlath im my view it will be overkill if we will try to write function to identify phone number around the world

**Review by puneetlath — CHANGES_REQUESTED** [review]
Mostly looks good to me. Just one comment.

File: src/CONST.js [hunk 2]
```diff
@@ -786,6 +786,7 @@ const CONST = {
         DIGITS_AND_PLUS: /^\+?[0-9]*$/,
         PHONE_E164_PLUS: /^\+?[1-9]\d{1,14}$/,
         PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]{0,12}$/,
```
[the remark this answers is not in the document]
Comment (narefyev91) [reply src/CONST.js]: @puneetlath fixed!

File: src/CONST.js [hunk 3]
```diff
@@ -785,7 +785,7 @@ const CONST = {
         US_PHONE_WITH_OPTIONAL_COUNTRY_CODE: /^(\+1)?\d{10}$/,
         DIGITS_AND_PLUS: /^\+?[0-9]*$/,
         PHONE_E164_PLUS: /^\+?[1-9]\d{1,14}$/,
-        PHONE_WITH_SPECIAL_CHARS: /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\\./0-9]{0,12}$/,
+        PHONE_WITH_ANY_CHARS: /^\s*(?:\+?(\d{1,3}))?[-. (]*(\d{3})[-. )]*(\d{3})[-. ]*(\d{4})(?: *x(\d+))?\s*$/,
```
Comment (puneetlath) [inline src/CONST.js:788]: Actually sorry, perhaps we should rename this to `PHONE_WITH_SPECIAL_CHARS`. Since it doesn't actually allow _any_ characters.

**Review by puneetlath — CHANGES_REQUESTED** [review]

File: src/CONST.js (hunk 3 above)
Comment (narefyev91) [reply src/CONST.js]: sure, no problem at all! done @puneetlath

**Review by puneetlath — APPROVED** [review]
Ok sweet. Looks good to me. @sobitneupane over to you!

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>

Screenshot 2023-03-02 at 23 37 37
Screenshot 2023-03-02 at 23 37 27
Screenshot 2023-03-02 at 23 37 15

</details>

<details>
<summary>Mobile Web - Chrome</summary>




</details>

<details>
<summary>Mobile Web - Safari</summary>





</details>

<details>
<summary>Desktop</summary>
Screenshot 2023-03-02 at 23 42 46
Screenshot 2023-03-02 at 23 42 36
Screenshot 2023-03-02 at 23 42 17


</details>

<details>
<summary>iOS</summary>





</details>
<details>
<summary>Android</summary>



</details>

**Comment by sobitneupane** [comment]
Video Comparison with and without special characters
| Without Special characters  | With Special Characters |
| ------------- | ------------- |
|<video src="https://user-images.githubusercontent.com/25876548/222519271-15e81026-bedb-4ae4-b288-dd61386927b7.mov" >| <video src="https://user-images.githubusercontent.com/25876548/222519283-62527a8d-f95a-49ca-b7b1-f4bc63f0eb47.mov">|

With Special Characters, contact is shown only after whole input is entered.

cc: @puneetlath

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Comment by sobitneupane** [comment]
@narefyev91 Can you please add [other formats](https://github.com/Expensify/App/pull/15594#issuecomment-1451794348) to be tested on Tests section in PR body?

**Comment by narefyev91** [comment]
> @narefyev91 Can you please add [other formats](https://github.com/Expensify/App/pull/15594#issuecomment-1451794348) to be tested on Tests section in PR body?

Added

**Comment by puneetlath** [comment]
>With Special Characters, contact is shown only after whole input is entered.

I think that's ok since it's not really a valid phone number until that point anyways.

**Pull request merged**